src/mpi_collectives/include/mpi_wrapper.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Copyright (c) 2018 by Contributors
+ */
+
+#ifndef MXNET_MPI_COLLECTIVES_INCLUDE_MPI_WRAPPER_H_
+#define MXNET_MPI_COLLECTIVES_INCLUDE_MPI_WRAPPER_H_
+
+#if MXNET_USE_MPI_DIST_KVSTORE
+
+#include <mpi.h>
+
+#include "mxnet/ndarray.h"
+#include "mxnet/base.h"
+#include "mpi_message.pb.h"
+
+template<typename DType>
+MPI_Datatype MPI_Data_Type_Cast(void) {
+  LOG(FATAL) << "Need to template specialization to get mpi data type";
```

Comment:
Why do you need a default definition that throw errors at runtime? You already added template specialization. Would a template definition like below be sufficient?
```
template<typename DType>
MPI_Datatype MPI_Data_Type_Cast(void);
```

Reply:
My original point is that if anyone used unsupported data type it can be caught by run time exception. In your way, it looks like any unsupported data type will be caught in compilation time. 
OK for me to remove default implementation.